storage/innobase/include/clone0clone.h

```diff
@@ -0,0 +1,1360 @@
+  @param[in]    func            callback function for condition check
+  @param[in]    mutex           release during sleep and re-acquire
+  @param[out]   is_timeout      true if timeout
+  @return error code returned by callback function. */
+  static int wait(Clone_Msec sleep_time, Clone_Sec timeout,
+                  Clone_Sec alert_interval, Wait_Cond_Cbk_Func &&func,
+                  mysql_mutex_t *mutex, bool &is_timeout) {
+    int err = 0;
+    bool wait = true;
+    is_timeout = false;
+
+    int loop_count = 0;
+    auto alert_count = static_cast<int>(alert_interval / sleep_time);
+    auto total_count = static_cast<int>(timeout / sleep_time);
+
+    /* Call function once before waiting. */
+    err = func(false, wait);
+
+    /* Start with 1 ms sleep and increase up to target sleep time. */
+    Clone_Msec cur_sleep_time{1};
+
+    while (!is_timeout && wait && err == 0) {
+      /* Release input mutex */
+      if (mutex != nullptr) {
+        mysql_mutex_assert_owner(mutex);
+        mysql_mutex_unlock(mutex);
+      }
+
+      /* Limit sleep time to what is passed by caller. */
+      if (cur_sleep_time > sleep_time) {
+        cur_sleep_time = sleep_time;
+      }
+
+      std::this_thread::sleep_for(cur_sleep_time);
+
+      if (cur_sleep_time < sleep_time) {
+        /* Double sleep time in each iteration till we reach target. */
+        cur_sleep_time *= 2;
+      } else {
+        /* Increment count once we have reached target sleep time. */
+        ++loop_count;
+      }
+
+      /* Acquire input mutex back */
+      if (mutex != nullptr) {
+        mysql_mutex_lock(mutex);
+      }
+
+      /* We have not yet reached the target sleep time. */
+      if (loop_count == 0) {
+        err = func(false, wait);
+        continue;
+      }
+
+      auto alert = (alert_count > 0) ? (loop_count % alert_count == 0) : true;
+
+      err = func(alert, wait);
+
+      is_timeout = (loop_count > total_count);
+    }
+    return (err);
+  }
+
+  /** Wait till the condition is satisfied or default timeout.
+  @param[in]    func            callback function for condition check
+  @param[in]    mutex           release during sleep and re-acquire
+  @param[out]   is_timeout      true if timeout
+  @return error code returned by callback function. */
+  static int wait_default(Wait_Cond_Cbk_Func &&func, mysql_mutex_t *mutex,
+                          bool &is_timeout) {
+    return (wait(CLONE_DEF_SLEEP, Clone_Sec(CLONE_DEF_TIMEOUT),
+                 CLONE_DEF_ALERT_INTERVAL,
+                 std::forward<Wait_Cond_Cbk_Func>(func), mutex, is_timeout));
+  }
+
+  /** Check if any active clone is running.
+  @param[in]    print_alert     print alert message
+  @return true, if concurrent clone in progress */
+  bool check_active_clone(bool print_alert);
+
+  /** Check if any active clone is running.
+  @return (true, handle) if concurrent clone in progress */
+  std::tuple<bool, Clone_Handle *> check_active_clone();
```

Comment:
Given that the first function is only being invoked with `print_alert=false`, we can remove this function altogether, and just invoke the more complex `check_active_clone()`, which must be revised:
```c++
/**
@return handle of active clone
@retval nullptr if no clone is in progress */
Clone_handle *check_active_clone();
```